### PR TITLE
Feature advertisement service

### PR DESCRIPTION
**Pull request by geniusivanrozanov**

### CODE REVIEW

File: Services/IdentityService/IdentityService.Application/DataInitializers/IdentityDataInitializer.cs [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+        await SeedUsersAsync(userManager);
+    }
+
+    private static async Task SeedRolesAsync(RoleManager<IdentityRole<Guid>> roleManager)
+    {
+        var roles = Roles.GetAllRoles();
+
+        foreach (var role in roles)
+        {
```
Comment (katyalys) [inline Services/IdentityService/IdentityService.Application/DataInitializers/IdentityDataInitializer.cs:21]: i guess you can remove loop from here

File: Services/IdentityService/IdentityService.Application/DataInitializers/IdentityDataInitializer.cs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+            var existingUser = await userManager.FindByEmailAsync(user.entity.Email!);
+            
+            if (existingUser is null)
+            {
+                var result = await userManager.CreateAsync(user.entity, user.password);
+                
+                if (!result.Succeeded)
+                {
+                    throw new ArgumentException(
+                        $"Failed to create user with errors: {string.Join(", ", result.Errors.Select(e => e.Description))}");
```
Comment (katyalys) [inline Services/IdentityService/IdentityService.Application/DataInitializers/IdentityDataInitializer.cs:67]: define class with constants with exception messages

File: Services/Advertisement/Advertisement.Infrastructure/Data/Repositories/AdRepository.cs [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+    public async Task<bool> ExistsWithVinAsync(string vin, CancellationToken cancellationToken = default)
+    {
+        return await _context.Ads
+            .Find(x => x.Vin == vin)
+            .AnyAsync(cancellationToken);
+    }
+
+    public async Task CreateAd(AdEntity entity)
+    {
```
Comment (katyalys) [inline Services/Advertisement/Advertisement.Infrastructure/Data/Repositories/AdRepository.cs:81]: why dont you pass cancellation token to this method?

File: Services/Advertisement/Advertisement.Infrastructure/Data/Repositories/AdRepository.cs [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        await _context.Ads
+            .InsertOneAsync(entity);
+    }
+
+    public async Task UpdateAd(AdEntity entity)
+    {
```
Comment (katyalys) [inline Services/Advertisement/Advertisement.Infrastructure/Data/Repositories/AdRepository.cs:87]: same

**Review by katyalys — CHANGES_REQUESTED** [review]

File: Services/IdentityService/IdentityService.Application/DataInitializers/IdentityDataInitializer.cs (hunk 1 above)
Comment (geniusivanrozanov) [reply Services/IdentityService/IdentityService.Application/DataInitializers/IdentityDataInitializer.cs:21]: Here i check the existence of all roles, that defined in constants

File: Services/IdentityService/IdentityService.Application/DataInitializers/IdentityDataInitializer.cs (hunk 2 above)
Comment (geniusivanrozanov) [reply Services/IdentityService/IdentityService.Application/DataInitializers/IdentityDataInitializer.cs:67]: different messages are generated in different cases

File: Services/Advertisement/Advertisement.Application/Features/Commands/CreateAd/CreateAdCommandHandler.cs [hunk 5]
```diff
@@ -53,7 +53,7 @@ public async Task<GetAdDto> Handle(CreateAdCommand request, CancellationToken ca
 
         entity.OwnerId = _user.Id;
```
Comment (mebry) [inline Services/Advertisement/Advertisement.Application/Features/Commands/CreateAd/CreateAdCommandHandler.cs:55]: For asynchronous methods, it's necessary to append Async to the end of the method name

File: Services/Advertisement/Advertisement.Application/Features/Commands/DeleteAd/DeleteAdCommandHandler.cs [hunk 6]
```diff
@@ -35,6 +35,6 @@ public async Task Handle(DeleteAdCommand request, CancellationToken cancellation
         entity.UpdatedAt = currentTime;
         entity.Status = AdStatus.Deleted;
 
-        await _adRepository.UpdateAd(entity);
+        await _adRepository.UpdateAd(entity, cancellationToken);
```
Comment (mebry) [inline Services/Advertisement/Advertisement.Application/Features/Commands/DeleteAd/DeleteAdCommandHandler.cs:38]: The same

File: Services/Advertisement/Advertisement.WebAPI/Services/CurrentUser.cs [hunk 7]
```diff
@@ -14,7 +14,5 @@ public CurrentUser(IHttpContextAccessor httpContextAccessor)
     }
 
     public Guid Id => Guid.Parse(_user.FindFirstValue(ClaimTypes.NameIdentifier)!);
-    public bool IsUser => _user.IsInRole(Roles.User);
-    public bool IsAdmin => _user.IsInRole(Roles.Admin);
-    public bool IsModerator => _user.IsInRole(Roles.Moderator);
+    public bool IsInRole(string role) => _user.IsInRole(role);
```
Comment (mebry) [inline Services/Advertisement/Advertisement.WebAPI/Services/CurrentUser.cs:17]: don't use =>. Instead, explicitly use curly braces, which will store the logic of the method, even if there is 1 line

File: Services/IdentityService/IdentityService.Application/DataInitializers/IdentityDataInitializer.cs (hunk 2 above)
Comment (mebry) [reply Services/IdentityService/IdentityService.Application/DataInitializers/IdentityDataInitializer.cs:67]: Yes, you're right that you generate different text for different errors, but it says here that you need to create a separate class in which all messages will be stored so as not to write them explicitly here

File: Services/Advertisement/Advertisement.WebAPI/Services/CurrentUser.cs [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+{
+    private readonly ClaimsPrincipal _user;
+
+    public CurrentUser(IHttpContextAccessor httpContextAccessor)
+    {
+        _user = httpContextAccessor.HttpContext!.User;
+    }
+
+    public Guid Id => Guid.Parse(_user.FindFirstValue(ClaimTypes.NameIdentifier)!);
+    public bool IsInRole(string role) => _user.IsInRole(role);
```
Comment (mebry) [inline Services/Advertisement/Advertisement.WebAPI/Services/CurrentUser.cs:17]: When describing a class, use the following structure:
· Declare all class members, including readonly or constants
· Declare all public properties, including auto-properties
· Declare all constructor(s)
· Declare the destructor, if any
· Declare all public methods
· Declare all protected methods
· Declare all private methods

File: Services/IdentityService/IdentityService.Application/DataInitializers/IdentityDataInitializer.cs (hunk 1 above)
Comment (mebry) [reply Services/IdentityService/IdentityService.Application/DataInitializers/IdentityDataInitializer.cs:21]: It makes sense

> Here i check the existence of all roles, that defined in constants

**Review by mebry — CHANGES_REQUESTED** [review]

**Review by mebry — APPROVED** [review]

**Pull request merged**